src/compute-client/src/controller.rs

```diff
@@ -562,7 +562,7 @@ where
     }
 
     /// Processes the work queued by [`ComputeController::ready`].
-    pub fn process(&mut self) -> Option<ComputeControllerResponse<T>> {
+    pub async fn process(&mut self) -> Option<ComputeControllerResponse<T>> {
```

Comment:
I'm a bit concerned about this change. Usually in the `ready`/`process` pattern, we avoid running anything in `process` that can block or take a long time. Now, this can potentially block in `append_to_collection`, if the update channel's queue is full: https://github.com/MaterializeInc/materialize/blob/2bd546502e/src/storage-controller/src/collection_mgmt.rs#L148-L148

Maybe it is fine, but we should check carefully that nothing will break if this function blocks for a while.

Reply:
Two ways to avoid this entirely could be (1) firing off a task to drive the future, or (2) making the channels that `append_to_collection` uses be unbounded.

Reply:
btw, I guess the size of the channel is 255: https://github.com/MaterializeInc/materialize/blob/51a54bd1e8/src/storage-controller/src/collection_mgmt.rs#L33-L34

so for this to block, there would need to be 255 outstanding writes to one of these tables, which seems hard to imagine (unless everything is totally broken anyway).

Reply:
>  so it's probably fine to make this async.

Yes, but there's a distinction between things that just have to be async because of the function-color problem, and things that are "meaningfully" async; that is, that might really block for a long time. I think the former is okay in `process` functions (though it might be clearer if we made `process` functions sync and just used `block_on` within them), whereas the second is legitimately problematic.

Reply:
> whereas the second is legitimately problematic.

This addition to `process` would fall into the second category? I'm also just now noticing the comment on this method which says:

>  /// This method is guaranteed to return "quickly" unless doing so would 
>  /// compromise the correctness of the system. 

Which would no longer seem true if we merged this. I would be more comfortable if we kept this method as sync. Is there any issues with one of the suggestions from @umanwizard?

Reply:
> This addition to `process` would fall into the second category?

That's what is a bit unclear to me. It _can_, but only if  the channel in `append_to_collection` is full, which can only happen if there are 255 outstanding writes to the collection, which seems unlikely.

@ParkMyCar , is there any reason why the channels that function uses can't/shouldn't be made unbounded?

Reply:
That channel is bounded to provide back pressure to writers. It's _probably_ fine if we want to make it unbounded, but theoretically it could cause memory to blow up and thus OOM environmentd

Reply:
If the method is sync and uses unbounded channels that surpasses some amount of messages (> 255) then environmentd will OOM which is bad. If the method is async and uses bounded channels that surpasses 255 messages, then the main coord loop will block which is bad. So both approaches don't seem great. What about the other suggestion?

> (1) firing off a task to drive the future

Reply:
That task would have to consolidate staged introspection data when it can't write, otherwise it would still produce an OOM because the data needs to be kept _somewhere_. This gets more complex.

We already call `append_to_managed_collections` at several places in the storage controller's `process` method ([for example](https://github.com/MaterializeInc/materialize/blob/d59e02ef6c0482435a3b535022d656978b167e21/src/storage-controller/src/lib.rs#L1638)), so calling it in the compute controller's doesn't make things worse at least. I don't see a reason to worry to much about this instance. It might be worth taking a good look at all the `await` calls that happen in the coordinator's message loop currently, but that's outside the scope of this PR.

Reply:
> I don't see a reason to worry to much about this instance.

I agree.

Reply:
Well, it says it will return "quickly", so there is some room for interpretation :D I think we do want the method to return quickly and the question is just whether the current `await` calls fulfill that requirement. So it's probably a good idea to leave that part in the comment, so we are reminded. I agree that we should probably remove the "guaranteed" part because there is nothing guaranteeing a quick return.